### PR TITLE
Introduce extend api catalog method

### PR DESCRIPTION
**Pull request by LironMShemen**

### CODE REVIEW

File: src/test-utils/expendAPICatalogUtils.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        {
+            signature: 'swipe(direction: string)',
+            description: 'Swipes in the specified direction.',
+            example: 'await swipe("up");',
+            guidelines: ['Use this method to scroll the screen.']
+        }
+    ]
+}];
+
+export const actionsCategory = [{
```
[the remark this answers is not in the document]
Comment (LironMShemen) [reply src/test-utils/expendAPICatalogUtils.ts:13]: You're right. Initially, I wanted to allow the user to add more than one category, so I let them insert a list of categories. I have now added the option to insert either a single Category object or a list of Categories.

File: src/Copilot.test.ts [hunk 2]
```diff
@@ -188,4 +194,43 @@ describe('Copilot', () => {
...
+        });
+
+        it('should expend the API catalog with an existing category and an additional new category', () => {
+            Copilot.init(mockConfig);
+            const instance = Copilot.getInstance();
+
+            // item is added to existing category
+            instance.extendAPICatalog(actionsCategory)
+            instance.extendAPICatalog(actionsCategory2, tapButtonContext);
+            expect(mockConfig.frameworkDriver.apiCatalog.categories).toEqual([...actionsCategory]);
```
[the remark this answers is not in the document]
Comment (LironMShemen) [reply src/Copilot.test.ts:227]: I think this is exactly what happens :)
You can see in the utils file that `actionCategory` and `actionCategory2` have the same title, `Actions`.

File: src/Copilot.ts [hunk 3]
```diff
@@ -21,8 +21,10 @@ export class Copilot {
     private stepPerformer: StepPerformer;
     private cacheHandler: CacheHandler;
     private isRunning: boolean = false;
+    private config: Config;
 
     private constructor(config: Config) {
+        this.config = config;
```
[the remark this answers is not in the document]
Comment (LironMShemen) [reply src/Copilot.ts:27]: I don’t need it :)
 Maybe I thought of it at the beginning. 
Removed.

File: src/Copilot.test.ts [hunk 4]
```diff
@@ -1,9 +1,14 @@
...
-import { mockCache, mockedCacheFile } from "./test-utils/cache";
+import {Copilot} from '@/Copilot';
+import {StepPerformer} from '@/actions/StepPerformer';
+import {CopilotError} from '@/errors/CopilotError';
+import {Config} from "@/types";
+import {mockCache, mockedCacheFile} from "./test-utils/cache";
+import {
+    customActionsCategory,
+    actionsCategory2,
+    actionsCategory,
```
Comment (asafkorem) [inline src/Copilot.test.ts:9]: naming is confusing (both refer to actions)
use `barCategory` and `bazCategory1`, `bazCategory2` or something? just not to mention actions in both categories

File: src/Copilot.ts [hunk 5]
```diff
@@ -109,6 +109,17 @@ export class Copilot {
             this.cacheHandler.flushTemporaryCache();
     }
 
+    /**
+     * Allow the user to add context and categories (titles and items) to the existing API catalog.
```
Comment (asafkorem) [inline src/Copilot.ts:113]: Enriches the API catalog with given categories and JS context.

Comment (asafkorem) [reply src/Copilot.ts:113]: or "Adds ...". 
the description should describe what the method does (its effect)

File: src/actions/StepPerformer.test.ts [hunk 6]
```diff
@@ -283,4 +284,38 @@ describe('StepPerformer', () => {
...
+        it('should extend the context with the given object', async () => {
+            // Initial context
+            stepPerformer.extendJSContext(dummyBarContext1);
+
+            setupMocks();
+            await stepPerformer.perform(INTENT);
+            expect(mockCodeEvaluator.evaluate).toHaveBeenCalledWith(PROMPT_RESULT, dummyBarContext1);
+
+            // Extended context
+            const newFixedContext = { ...dummyBarContext1, ...dummyContext };
```
Comment (asafkorem) [inline src/actions/StepPerformer.test.ts:298]: I think `expectedContext` or `extendedContext` is more semantically correct here

File: src/actions/StepPerformer.ts [hunk 7]
```diff
@@ -19,9 +19,19 @@ export class StepPerformer {
...
 
+    extendJSContext(newContext: any): void {
+        for (const key in newContext) {
+            if (key in this.context) {
+                console.log(`Notice: Context ${key} is overridden by the new context value`);
+                break;
+            }
+        }
+        this.context = {...this.context, ...newContext};
+    }
```
Comment (asafkorem) [inline src/actions/StepPerformer.ts:30]:
```ts
extendJSContext(newContext: any): void {
  const overriddenKey = Object.keys(newContext).find(key => key in this.context);
  
  if (overriddenKey) {
    console.log(`Notice: Context ${overriddenKey} is overridden by the new context value`);
  }
  
  this.context = {
    ...this.context,
    ...newContext,
  };
}
```

Comment (asafkorem) [reply src/actions/StepPerformer.ts:30]: Thinking about this now, we're not merging categories on the first initialization if user passes two categories with the same name.
It makes sense to me if we'll do the same thing there. But IDK, kind of an edge case that doesn't bother me much.

File: src/integration tests/index.test.ts [hunk 8]
```diff
@@ -330,4 +332,27 @@ describe('Copilot Integration Tests', () => {
...
+
+        beforeEach(() => {
+            jest.clearAllMocks();
+            copilot.init({
+                frameworkDriver: mockFrameworkDriver,
+                promptHandler: mockPromptHandler,
+            });
+            copilot.start();
+        });
+        it('should call relevant functions to extend the catalog', () => {
```
Comment (asafkorem) [inline src/integration tests/index.test.ts:348]: missing empty line

File: src/types.ts [hunk 9]
```diff
@@ -41,6 +41,14 @@ export interface CopilotFacade {
...
      */
     perform: (...steps: string[]) => Promise<string>;
+
+    /**
+     * Extends the API catalog of the testing framework with additional categories and items.
+     * @param context The variables of the testing framework (i.e. exposes the matching function, expect, etc.).
+     * @param categories The categories to add to the API catalog.
+     * @note This can be used to add custom categories and items to the API catalog.
+     */
+    extendAPICatalog: (categories: TestingFrameworkAPICatalogCategory[] | TestingFrameworkAPICatalogCategory, context?: any,) => void;
```
Comment (asafkorem) [inline src/types.ts:51]: PERFECT

Comment (asafkorem) [reply src/types.ts:51]: `categories and items` → `APIs (categories and JS context)`

File: src/utils/PromptCreator.test.ts [hunk 10]
```diff
@@ -93,4 +94,23 @@ describe('PromptCreator', () => {
 
         expect(prompt).toMatchSnapshot();
     });
+
+    describe('extentAPICategories', () => {
+        it('should extend the API catalog with new or exist category', () => {
+            // Add new category
```
Comment (asafkorem) [inline src/utils/PromptCreator.test.ts:100]: test code should be semantic enough so the comment will be redundant

File: src/utils/PromptCreator.ts [hunk 11]
```diff
@@ -6,7 +6,22 @@ import {
...
+            const existingCategory = this.apiCatalog.categories.find((existingCategory) => existingCategory.title === category.title);
+            if (existingCategory) {
+                for (const item of category.items) {
+                    existingCategory.items.push(item);
+                }
+            } else {
+                this.apiCatalog.categories.push(category);
+            }
+        }
+    }
```
Comment (asafkorem) [inline src/utils/PromptCreator.ts:24]: 1. too many indentations (third-level blocks). let's try to flatten the code:


```ts
extendAPICategories(
  newCategories: TestingFrameworkAPICatalogCategory[] | TestingFrameworkAPICatalogCategory
): void {
  const categories = Array.isArray(newCategories) ? newCategories : [newCategories];

  categories.forEach((category) => {
    const existingCategory = this.apiCatalog.categories.find(
      (c) => c.title === category.title
    );

    if (existingCategory) {
      existingCategory.items.push(...category.items);
    } else {
      this.apiCatalog.categories.push(category);
    }
  });
}
```

Comment (asafkorem) [reply src/utils/PromptCreator.ts:24]: 2. I think it's better if we won't allow passing a single category as an argument for parameter that is called `categories`

**Review by asafkorem — CHANGES_REQUESTED** [review]
good job. round 2

File: src/Copilot.test.ts [hunk 12]
```diff
@@ -1,9 +1,14 @@
...
+    tapButtonContext
+} from "./test-utils/expendAPICatalogUtils";
```
[the remark this answers is not in the document]
Comment (asafkorem) [reply src/Copilot.test.ts]: done

File: src/test-utils/extendAPICatalogUtils.ts [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+            guidelines: ['Use this method to scroll the screen.']
+        }
+    ]
+};
+
+export const dummyContext = {foo: jest.fn()};
+export const dummyBarContext1 = {bar: jest.fn()};
+export const dummyBarContext2 = {bar: jest.fn()};
+
+export const promptCreatorConstructorMockAPI: TestingFrameworkAPICatalog = {
```
Comment (asafkorem) [inline src/test-utils/extendAPICatalogUtils.ts:43]: it shouldn't be under `extendAPICatalogUtils` because you're now using it for the constructor, let's just call the file `APICatalogTestUtils.ts`

File: src/utils/PromptCreator.ts [hunk 14]
```diff
@@ -6,7 +6,36 @@ import {
 } from "@/types";
 
 export class PromptCreator {
-    constructor(private apiCatalog: TestingFrameworkAPICatalog) {}
+    constructor(private apiCatalog: TestingFrameworkAPICatalog) {
```
Comment (asafkorem) [inline src/utils/PromptCreator.ts:9]: Both the constructor and `extendAPICategories` are doing the same operation - merging categories based on their titles. We can extract this shared logic and use it within one util method.

Comment (asafkorem) [reply src/utils/PromptCreator.ts:9]: e.g.

```ts
    private mergeCategories(categories: TestingFrameworkAPICatalogCategory[]): void {
        const categoryMap = this.apiCatalog.categories.reduce((acc, category) => {
            acc.set(category.title, category);
            return acc;
        }, new Map<string, TestingFrameworkAPICatalogCategory>());

        categories.forEach(category => {
            const existing = categoryMap.get(category.title);
            if (existing) {
                existing.items.push(...category.items);
            } else {
                categoryMap.set(category.title, {...category});
            }
        });

        this.apiCatalog.categories = Array.from(categoryMap.values());
    }
```

Comment (asafkorem) [reply src/utils/PromptCreator.ts:9]: also we can use more functional pattern - [reduce](https://geeksforgeeks.org/typescript-array-reduce-method/):
```ts
private mergeCategories(categories: TestingFrameworkAPICatalogCategory[]): TestingFrameworkAPICatalogCategory[] {
    return categories.reduce((mergedCategories, category) => {
        const existingIndex = mergedCategories.findIndex(c => c.title === category.title);
        
        if (existingIndex >= 0) {
            mergedCategories[existingIndex].items.push(...category.items);
            return mergedCategories;
        }

        return [...mergedCategories, {...category}];
    }, [] as TestingFrameworkAPICatalogCategory[]);
}
```

Comment (asafkorem) [reply src/utils/PromptCreator.ts:9]: if category doenst exist then existingIndex will be -1 (that's how findIndex works)

Comment (asafkorem) [reply src/utils/PromptCreator.ts:9]: then, we can call the same method this way:

```ts
constructor(private apiCatalog: TestingFrameworkAPICatalog) {
    this.mergeCategories(this.apiCatalog.categories);
}

extendAPICategories(
    newCategories: TestingFrameworkAPICatalogCategory[]
): void {
    this.apiCatalog.categories = this.mergeCategories([...this.apiCatalog.categories, ...newCategories]);
}
```

Comment (asafkorem) [reply src/utils/PromptCreator.ts:9]: also I removed the option of one category from my example as we said we'll change it to only array

**Review by asafkorem — CHANGES_REQUESTED** [review]
Last round!! 🙂 
Amazing work so far 🏆

File: src/utils/PromptCreator.ts (hunk 14 above)
Comment (asafkorem) [reply src/utils/PromptCreator.ts:9]: Now, that we have a shared logic in a common function - we can maintain it from one place and improve it without duplicating.
For example - what happens if someone accidently calls the extend function with the same parameters over and over again (e.g. decided to put that on the `beforeEach` which is called multiple time)? We want to avoid redundancy and duplications in the new catalog. 

So we can warn / err them or just avoid the duplications, to do that, we can use the old trick of transforming the array of items to set → back to array, e.g.:

```ts
const uniqueItems = Array.from(new Set(category.items));
```

File: src/utils/PromptCreator.ts [hunk 15]
```diff
@@ -6,7 +6,32 @@ import {
...
+        this.apiCatalog.categories = this.mergeCategories([...this.apiCatalog.categories, ...newCategories]);
+    }
+
+    private mergeCategories(categories: TestingFrameworkAPICatalogCategory[]): TestingFrameworkAPICatalogCategory[] {
+        return categories.reduce((mergedCategories, category) => {
+            const existingIndex = mergedCategories.findIndex(c => c.title === category.title);
+
+            if (existingIndex >= 0) {
+                const uniqueItems = [...mergedCategories[existingIndex].items, ...category.items];
+                mergedCategories[existingIndex].items = Array.from(new Set(uniqueItems));
```
Comment (asafkorem) [inline src/utils/PromptCreator.ts:25]: `uniqueItems` is not unique items, it's just merged items

File: src/utils/PromptCreator.ts [hunk 16]
```diff
@@ -6,7 +6,32 @@ import {
...
+                return mergedCategories;
+            } else {
+                const uniqueItems = Array.from(new Set(category.items));
+                category.items = uniqueItems;
```
Comment (asafkorem) [inline src/utils/PromptCreator.ts:29]:
```suggestion
                category.items = uniqueItems(category.items);
```

File: src/utils/PromptCreator.ts [hunk 17]
```diff
@@ -6,7 +6,32 @@ import {
...
+        newCategories: TestingFrameworkAPICatalogCategory[]
+    ): void {
+        this.apiCatalog.categories = this.mergeCategories([...this.apiCatalog.categories, ...newCategories]);
+    }
+
+    private mergeCategories(categories: TestingFrameworkAPICatalogCategory[]): TestingFrameworkAPICatalogCategory[] {
+        return categories.reduce((mergedCategories, category) => {
+            const existingIndex = mergedCategories.findIndex(c => c.title === category.title);
+
+            if (existingIndex >= 0) {
```
Comment (asafkorem) [inline src/utils/PromptCreator.ts:23]:
```suggestion

            const uniqueItems = (items) => Array.from(new Set(items));

            if (existingIndex >= 0) {
```

**Review by asafkorem — CHANGES_REQUESTED** [review]

**Review by asafkorem — APPROVED** [review]

**Comment by asafkorem** [comment]
@LironMShemen you da best!!!
